tchannel/tornado/connection.py

```diff
@@ -701,3 +634,176 @@ def _request_timed_out(self, req_id, req_ttl, future):
         # while.
         future.set_exception(errors.TimeoutError())
         self._request_tombstones.add(req_id, req_ttl)
+
+
+class Reader(object):
+
+    def __init__(self, io_stream, capacity=64):
+        self.queue = tornado.queues.Queue(capacity)
+        self.filling = False
+        self.io_stream = io_stream
+
+    def fill(self):
+        self.filling = True
+
+        io_loop = IOLoop.current()
+
+        def keep_reading(f):
+            if f.exception():
+                self.filling = False
+                return log(f.exception())
+            # connect these two in the case when put blocks
+            self.queue.put(f.result())
```

Comment:
If you don't wait on the future returned by `queue.put` before calling `fill` again, specifying a capacity when you construct the queue is useless. This won't slow down reading, it'll just keep queueing up futures in the IOLoop that will eventually `put` into the queue. You want to do something like this:

```python
self.queue.put(f.result()).add_done_callback(lambda f: io_loop.spawn_callback(self.fill))
# It's safe to use add_done_callback here because spawn_callback already delays the call
# until the next iteration of the IOLoop.
```

(Or also handle the exception for `queue.put`, although I'm not aware of any recoverable exceptions raised by `queue.put`.)